### ISSUE TITLE
Callbacks - how?

### ISSUE THREAD

**Issue by SzymonPobiega** [Particular/NServiceBus#3369]
This issue is meant to discuss possible approaches to callbacks. For now I can see two alternatives but feel free to add more.

## Callbacks in the transport

In this approach we implement callbacks on the transport level:
 * RabbitMQ uses [direct reply-to](https://www.rabbitmq.com/direct-reply-to.html)
 * SQL Server uses a single queue but has additional column for *callback id*. The receiver uses `WHERE CallbackID = <my_id> OR CallbackID IS NULL` when receiving. There is additional process that removes outdated callback messages if their destination endpoint instance has been decommissioned
 * ASB uses a spacial (possibly shared between endpoints) session queue for messages that require callbacks. Unfortunately this requires the receiving side to know about that queue (not backwards-compatible by default) -- @SeanFeldman please correct my if my thinking is wrong here
 * ASQ - ?
 * MSMQ - Works out of the box by means or `reply-to` header containing the sender's queue

### Pros

 * We can use native transport features
 * Does not affect the routing in any way

### Cons

 * Requires additional code to be written in all the transports
 * Requires merging back in the callbacks package because in order for transport to implement callbacks, the notion of a callback has to be part of the transport seam

## Callbacks based on routing

In this approach we use following definitions:
 * Endpoint refers to all the deployments of the same binary package (e.g. `sales`)
 * Logical instance refers to all deployments that share same *instance ID* (e.g. `sales-1`)
 * Physical instance refers to a single deployment
 * *Instance ID* is provides by the host

Callbacks just use the `reply-to` header. In RabbitMQ we would have to have some logical that would start two receivers if *instance ID* is not empty. The main receiver would not use the ID (taking advantage of competing consumers) and the other would be ID-specific (allowing callbacks)

### Pros

 * Allows to send commands to specific instances. Useful if there are multiple instances (same binaries) but are not part of the cluster (they are logically distinct) e.g. physical device controllers.
 * No additional code in transports
 * `NServiceBus.Callbacks` can probably remain a separate package as there is no additional information passed in the transport seam

### Cons

 * Requires more complex mental model of routing: even when users use broker transports, they might need to know what *instance ID* it. The assumption that it is, by default, provided by the host, makes this point slightly less important.
 * Generally requires more queues to be present in the underlying infrastructure (in ASP, SQL and RabbitMQ transports) - more stuff to care about.

**Comment by SzymonPobiega** [comment]
@Particular/tf-async-transport @johnsimons @bording @mauroservienti @tmasternak @SeanFeldman @yvesgoeleven please chime in. Apologies in advance if this sounds biased. Tried to be as neutral as I could but I know I failed. I believe that if we together put effort in finding and tearing apart all the pros and cons of both approaches (and possibly finding even better alternatives) the effect would be awesome. 

And I believe we can do it async, here, on github, so that everybody has equal chance to have his voice heard.

**Comment by johnsimons** [comment]
@SzymonPobiega regarding these cons:

1. Requires additional code to be written in all the transports
1. Requires merging back in the callbacks package because in order for transport to implement callbacks, the notion of a callback has to be part of the transport seam

Regarding 1, I actually see that as a pro, the less code we share the easier is to release out of band fixes, putting the shared code in the core it would mean that we locked in into yearly releases

Regarding 2, I am pretty sure that the way the callbacks are impl it uses a special header to tell the server that the client is waiting on a callback, so I would think that we can use that special header as well.

**Comment by johnsimons** [comment]
I am not sure where this fits, but @bording and I believe that callbacks queue should not be durable in Rabbit, but thinking further about it, I wonder if that can be applied to all transports?

So in Rabbit we are currently thinking, at startup we create a temporary (random queue name) for the callbacks, that queue is automatically deleted if the process dies. This is actually the way Rabbit does it natively, see https://www.rabbitmq.com/direct-reply-to.html
It is OK to delete these temporary queues because we would lose the callback anyway (in memory dictionary).
So in summary, Rabbit does not need to be aware of the “discriminator”, we will be handling the callbacks at the transport level and in scale out scenarios the main queue is shared (hence competing consumers).

**Comment by mauroservienti** [comment]
@johnsimons 

> I am not sure where this fits, but @bording and I believe that callbacks queue should not be durable in Rabbit, but thinking further about it, I wonder if that can be applied to all transports?

I like it, unfortunately there is no way to that in all transports, SQL for example has no concept of volatile queues, queues are managed by the endpoint and not by the broker so if the endpoint crashes there nothing we can do.

**Comment by johnsimons** [comment]
Another improvement/add on, all transports should set a TTL on the reply message because in reality the http timeout will kick in before the TTL, so we don't have to deal with clean ups since they are volatile messages.
So a TTL of 5 minutes on these messages would be plenty :smile:

**Comment by johnsimons** [comment]
@mauroservienti Yes I know, some transports unfortunately don't have the concept of non durable queues, hence the reason for https://github.com/Particular/NServiceBus/issues/3369#issuecomment-176017061.

But yes for those transports there is a bit of cleanup to do on these temp queues, but IMO a small price to pay.

**Comment by SeanFeldman** [comment]
To add to the TTL idea - we should tie that to the Web commonly accepted standard for request timeout. That way we can give our customers a number (time period) that we're promise to deliver a reply / know for how long to hold on to a callback registration.

**Comment by mauroservienti** [comment]
TTL should be there, I agree, we discussed it with @SeanFeldman as well yesterday.
BTW: I like the idea that the callback technical implementation can be different across transports, the business behavior must be the same, but if on a specific transport we can leverage specific features, such as non durable Qs in Rabbit, why not?

**Comment by johnsimons** [comment]
@SzymonPobiega I would be onboard with "Callbacks based on routing" if and only if, all that feature is external to the core so we are not locked into yearly major releases.

**Comment by mauroservienti** [comment]
that's why the idea is to try to spike callbacks using the actual separate package, so that we may discover if Core can ignore callbacks at all @johnsimons

**Comment by johnsimons** [comment]
what separate package ?
are you talking about NServiceBus.Callbacks ?

**Comment by mauroservienti** [comment]
yes

**Comment by johnsimons** [comment]
That package IMO is a different concern.
I see transport impls having to agree on a know contract and one of the points is to support callbacks

**Comment by mauroservienti** [comment]
yeah, I agree. Let me rephrase: the idea is to try to spike callbacks without involving Core

**Comment by johnsimons** [comment]
> the idea is to try to spike callbacks without involving Core

I like that @mauroservienti, but is it the right time to do that now, it sounds to me that such a feature can be released out of band

@SzymonPobiega what are your thoughts ?

**Comment by SzymonPobiega** [comment]
@mauroservienti @johnsimons we can try with SQL transport now. The only 'extra' effort I can see is adding that shared receiver in RabbitMQ which is necessary to maintain the `deliver once to one instance` semantics of an event. 

If we decide to go with callbacks in routing, core would know nothing about them:
 * The `NServiceBus.Callbacks` package would make sure to generate its own message id
 * It would register the continuation in a dictionary using that message id as a key
 * The `reply-to` header would include the endpoint's unique ID
 * The replier would copy the message id to `CorrelationId`
 * The callback sender would have a behavior (as it has now) injected before the handler execution which would inspect the `CorrelationId` and, if it finds a continuation under that id, would invoke that continuation and terminate the pipeline

**Comment by SzymonPobiega** [comment]
I see the TTL issue as orthogonal to the problem here. No matter which route we chose, we should ensure callback messages do have some form of TTL.

**Comment by mauroservienti** [comment]
> we can try with SQL transport now

@tmasternak can you chime in and provide a status update on how callbacks are implemented right now in SQL? cheers.

**Comment by johnsimons** [comment]
> I see the TTL issue as orthogonal to the problem here

:+1: , was just me ensuring we do it

**Comment by johnsimons** [comment]
> The only 'extra' effort I can see is adding that shared receiver in RabbitMQ

Not following @SzymonPobiega

**Comment by johnsimons** [comment]
@SzymonPobiega I am pretty sure the NServiceBus.Callbacks package already does:

- It would register the continuation in a dictionary using that message id as a key
- The replier would copy the message id to CorrelationId
- The callback sender would have a behavior (as it has now) injected before the handler execution which would inspect the CorrelationId and, if it finds a continuation under that id, would invoke that continuation and terminate the pipeline

**Comment by SzymonPobiega** [comment]
@johnsimons than SQL callbacks should work in SQL transport without SQL transport knowing anything about them, just by means of assigning unique endpoint IDs.

The problem is with RabbitMQ, because if you assign endpoint a unique ID (a.k.a. discriminator), the events would get delivered to all the individual queues bound to event exchange. That's why if we go with callbacks via routing, we essentially say RabbitMQ has to support unique endpoint IDs and hence we need to add a shared queue that would be bound to event exchanges.

**Comment by johnsimons** [comment]
@SzymonPobiega we are proposing to randomly create a temp queue per endpoint instance

**Comment by SzymonPobiega** [comment]
@johnsimons yes, that would be for transport-implemented callbacks. In routing-implemented callbacks there per-instance queues would not be random, but based on the instance ID.

**Comment by tmasternak** [comment]
@mauroservienti Currently in v3 there is no callback specific logic in SqlTransport that is there to support callbacks. In v2 there was a dedicated queue/receiver to handle replies for callbacks.

When it comes to temporary queues in SQL we do not have such mechanisms but we could try approach @SzymonPobiega in the issue but that might not work with the older endpoints. 

In addition I would like to add another approach to discussion
We treat NServiceBus.Callbacks as a 'integration' for request-response scenarios and nothing else. It won't share any configuration with any other endpoint running in the same process. New endpoint instance which will have separate queue with proper TTL on the messages. 
Cons:
 * It will not support shared pipeline or container.
 * We would have to come-up with a reasonable scheme for naming callback-only endpoints   
Pros:
 * Transports and core do not need to now anything about callbacks.

**Comment by SzymonPobiega** [comment]
Here's a brain dump from my conversation with @johnsimons:

### Stuff we agree on

 * RabbitMQ (and probably SQL Server) should implement callbacks by using two queues/receivers. 
 * One receiver would be the main queue (taking advantage of competing consumers) 
 * The other receiver would have a random name. That random name would be generated on install time by the installer in `NServiceBus.Callbacks` package.
 * If you don't reference `NServiceBus.Callbacks`, there would be no additional queue/receiver

### Stuff we not yet agree on
 * How that random name is communicated to the transport:
   * @johnsimons (correct me please if I messed up): the transport should get that ID from the callbacks package, not involving the core and that random name should not be used for scale out/round-robin distribution
   * @szymonpobiega: we should use the *instance ID* concept we have in core (under the name *discriminator*) and `NServiceBus.Callbacks` should hook up to the core and provide its stable-random way of generating these IDs. The same API would be used by the Azure host to provide *role id* as *instance ID*.

### Outcomes
 * RabbitMQ is not blocked. The only thing that can change is the way the random thing is provided
 * SQL Server transport should be tested with the callbacks package to validate our assumptions
 * @szymonpobiega will spike the alternative *instance ID* API

**Comment by mauroservienti** [comment]
If I recall correctly another requirement we discussed yesterday is that
the main queue and the callback one need to share a single pipeline for
incoming messages. If, again I understood correctly, this should be OK out
of the box because a receiver is the thing at he transport level that pull
messages from the input queue and delivers them to the pipeline.

Correct?

On Thu, 28 Jan 2016 at 10:28, Szymon Pobiega <notifications@github.com>
wrote:

> Here's a brain dump from my conversation with @johnsimons
> <https://github.com/johnsimons>:
> Stuff we agree on
>
>    - RabbitMQ (and probably SQL Server) should implement callbacks by
>    using two queues/receivers.
>    - One receiver would be the main queue (taking advantage of competing
>    consumers)
>    - The other receiver would have a random name. That random name would
>    be generated on install time by the installer in NServiceBus.Callbacks
>    package.
>    - If you don't reference NServiceBus.Callbacks, there would be no
>    additional queue/receiver
>
> Stuff we not yet agree on
>
>    - How that random name is communicated to the transport:
>       - @johnsimons <https://github.com/johnsimons> (correct me please if
>       I messed up): the transport should get that ID from the callbacks package,
>       not involving the core and that random name should not be used for scale
>       out/round-robin distribution
>       - @szymonpobiega <https://github.com/szymonpobiega>: we should use
>       the *instance ID* concept we have in core (under the name
>       *discriminator*) and NServiceBus.Callbacks should hook up to the
>       core and provide its stable-random way of generating these IDs. The same
>       API would be used by the Azure host to provide *role id* as *instance
>       ID*.
>
> Outcomes
>
>    - RabbitMQ is not blocked. The only thing that can change is the way
>    the random thing is provided
>    - SQL Server transport should be tested with the callbacks package to
>    validate our assumptions
>    - @szymonpobiega <https://github.com/szymonpobiega> will spike the
>    alternative *instance ID* API
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/issues/3369#issuecomment-176081063>
> .
>
-- 
Mauro Servienti | Solution Architect | Particular Software

**Comment by SzymonPobiega** [comment]
Another brain dump, this time from a call with @tmasternak 

 * Idea: because of callbacks, we need to address each instances anyway. If so, why not just make that instance-specific addressing a first class citizen in the core?
 * The core always uses the endpoint name (without instance id) as the main receiver and starts the message pump passing the address
 * If either a user or a plugin (such as callbacks) provided instance id, *the core* spins up another receiver, passing an instance-specific address and attaches it to the same pipeline
 * The transport does not have to know about callbacks or even additional receivers. All it needs to do is start receiving from a queue
 * The non-instance-specific address is treated as the main one in the core. That means `ReplyTo` will by default go to the shared queue
 * We add `ReplyToMe()` extension to `SendOptions` which will tell the routing to put the instance-specific address in the `reply-to` header (assuming instance id is provided) or throw (if not provided). 
 * Given we have a cluster with instances that have unique IDs, the way we send messages to that cluster is fully dependent on the senders:
   * If a sender routing does not specify instance mapping, the dispatcher will send messages to the shared queue (taking advantage of competing consumers)
   * If a sender routing does specify instance mapping, the dispatcher will round-robin between instances

We could not find any holes in this thinking. It allows to keep transports super-simple and it does not add any additional complexity to the core compared to the current state in V6. Can you please try to challenge it (I know I am biased)?

**Comment by danielmarbach** [comment]
Like it. The only thing I didn't understand in your explanation is the
ReplyToMe. can you explain more?

**Comment by mauroservienti** [comment]
Weren't you the one that hates meetings? ;-)

On Thu, Jan 28, 2016, 15:31 Szymon Pobiega <notifications@github.com> wrote:

> Another brain dump, this time from a call with @tmasternak
> <https://github.com/tmasternak>
>
>    - Idea: because of callbacks, we need to address each instances
>    anyway. If so, why not just make that instance-specific addressing a first
>    class citizen in the core?
>    - The core always uses the endpoint name (without instance id) as the
>    main receiver and starts the message pump passing the address
>    - If either a user or a plugin (such as callbacks) provided instance
>    id, *the core* spins up another receiver, passing an instance-specific
>    address and attaches it to the same pipeline
>    - The transport does not have to know about callbacks or even
>    additional receivers. All it needs to do is start receiving from a queue
>    - The non-instance-specific address is treated as the main one in the
>    core. That means ReplyTo will by default go to the shared queue
>    - We add ReplyToMe() extension to SendOptions which will tell the
>    routing to put the instance-specific address in the reply-to header
>    (assuming instance id is provided) or throw (if not provided).
>    - Given we have a cluster with instances that have unique IDs, the way
>    we send messages to that cluster is fully dependent on the senders:
>       - If a sender routing does not specify instance mapping, the
>       dispatcher will send messages to the shared queue (taking advantage of
>       competing consumers)
>       - If a sender routing does specify instance mapping, the dispatcher
>       will round-robin between instances
>
> We could not find any holes in this thinking. It allows to keep transports
> super-simple and it does not add any additional complexity to the core
> compared to the current state in V6. Can you please try to challenge it (I
> know I am biased)?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/issues/3369#issuecomment-176209241>
> .
>
-- 
Mauro Servienti | Solution Architect | Particular Software

**Comment by tmasternak** [comment]
@danielmarbach ReplyToMe is a way to request that the reply will be sent to this specific endpointId. This is a basic facility that callbacks would use and provide correlation of response-to-continuation on top of that.

I think that we might still consider transport specific support for instance specific addressing. Either way we probably should start spiking that.

**Comment by andreasohlund** [comment]
>The transport does not have to know about callbacks or even additional receivers. All it needs to do is start receiving from a queue

I like it, the transports beeing aware of callbacks is a smell to me

**Comment by mauroservienti** [comment]
@andreasohlund as discussed in the transports sync there are advantages in allowing, allowing not making, the transport to be aware of callbacks, this would allow transports to opt-in and provide their own implementation of callbacks, or some part of it.

On Azure ServiceBus this means, as suggested by @SeanFeldman, that we can support callbacks using 2 queues only, one main Q where instances behave as expected in a competing consumers manner and another one with sessions enabled that allow to reply to a specific instance.
Other approaches could be used on RabbitMQ as well that can create non-durable queues that exists for the lifetime of the client process. Etc.

**Comment by bording** [comment]
@SzymonPobiega Your latest proposal sounds like the best approach suggested so far. I'm thinking that the benefit of the code streamlining this would bring to the transports outweighs the benefits of allowing for transport-specific optimizations (for example a transient callbacks queue in RabbitMQ).

How would this change interact with satellites? I wouldn't want to end up with a bunch of instance-specific Timeouts queues in addition to the non-instance-specific ones already created.

**Comment by johnsimons** [comment]
Rabbit also supports a single receiver that dequeues messages from multiple queues, that is actually the way we implemented the callbacks, this latest proposal would not suit us.

Also confused about the need for `ReplyToMe`, can you give a code example?

**Comment by SeanFeldman** [comment]
Trying to visualize the last brain dump ([here](https://github.com/Particular/NServiceBus/issues/3369#issuecomment-176209241))  

If going that route, I'd like to be able to auto-delete _instance specific_ queues after certain period of time of inactivity (ASB and Rabbit support that) for scale-out/scale-in scenarios. That means we'll need to identify upon queue creation if we're dealing with _instance specific_ queue or a _shared_ queue.

![image](https://cloud.githubusercontent.com/assets/1309622/12660260/1d1536ec-c5d0-11e5-9910-b573b8592ded.png)

Same need would be in place if I'd implement a single _callbacks_ queue for ASB to use with sessions.

![image](https://cloud.githubusercontent.com/assets/1309622/12660079/2ccd0228-c5cf-11e5-99bc-1a86dfd969cb.png)

**Comment by bording** [comment]
@johnsimons While we do currently let a single RabbitMQ consumer handle both queues, I'd be totally ok with going with @SzymonPobiega's current approach and letting the transport not even care about manually maintaining a separate callbacks queue. Everything about the implementation becomes simpler and we still get the functionality we need.

After talking with @SeanFeldman about the approach, it does look like it would work for both RabbitMQ and ASB. However, as @SeanFeldman mentioned, with an additional tweak in two places, the current proposal would allow us to optionally take advantage of some transport-specific optimizations (sessions for ASB, auto-delete queues for both). To be able to allow for that, the `QueueBindings` class that gets passed into `ICreateQueues.CreateQueueIfNecessary()` would need to be modified to pass in a bool that specifies if the `ReceivingAddress` is instance-specific or not. We'd also like to have that same bool included as part of the `PushSettings` class that gets passed into `IPushMessages.Init()`.

How feasible would it be to make those two additions?

**Comment by johnsimons** [comment]
> Idea: because of callbacks, we need to address each instances anyway. If so, why not just make that instance-specific addressing a first class citizen in the core?

Because this is just a feature  to help people get started, eventually we want them not to use it.
The same way we pulled out callbacks from the core, this IMO does not belong in there either, especially since we only adding it for the callbacks.

**Comment by SzymonPobiega** [comment]
@johnsimons the instance specific routing is not only callbacks. Here are some of the scenarios we could support. Neither of them is mainstream. All are rather advanced but I think worth the effort especially that the core has been already written to support the round-robin distribution
 * broadcasting a command to multiple receivers (e.g. for integration purposes
 * round-robin distribution between instances
 * sending messages via two brokers/namespaces for HA scenarios
 * overcoming single queue throughput limitations by using round-robin distribution
I'd call YAGNI on that but I believe we can do better. We have enough knowledge of message-driven systems in house to make an informed call.

**Comment by SzymonPobiega** [comment]
@bording I think that `InstanceSpecific` flag makes sense, with one caveat though. I don't think we should treat *instance specific* as *volatile* because of some of the scenarios I've outlined above. But I think a RabbitMQ-specific flag `MakeInstanceSpcificQueuesVolatile` would be awesome.

Regarding satellites, they would always be shared (no instance id).

Regarding receiving from multiple queues with one receiver I am open for discussion. Currently one queue equals one `IPushMessages` so the satellites get their own pumps. An alternative approach would be something like:

```
public interface IPushMessages
{
   Task Bind(Func<PushContext, Task> pipe, CriticalError criticalError, PushSettings settings);
}
```

which could be called multiple times, allowing the transport to decide if it wants one or multiple receivers.

**Comment by tmasternak** [comment]
@bording additional parameter on the QueueBinding would be enough to make sure that SubscriptionManager works on the proper input queue right?

**Comment by mauroservienti** [comment]
I tagged this as in scope for V6

**Comment by andreasohlund** [comment]
> But I think a RabbitMQ-specific flag MakeInstanceSpcificQueuesVolatile would be awesome.

Just so I understand, volatile queues would only be allowed for responses where there are only callbacks waiting? 

Beacuse if the user has a handler as well as a callback a volatile queue would essentially mean message loss? (since the handler might never fire if we remove the queue prematurely)

Is it valid to combine handlers and callbacks for the same response? (we allow it currently right?)

**Comment by tmasternak** [comment]
@andreasohlund we allow it currently. The question is do we still want to do it - is it really important. We were discussing that callback responses are inherently transient and it probably makes sense to have TTL on them either way.

**Comment by johnsimons** [comment]
What Tomasz said, I also think that now that we have the new callbacks api
that provides much richer response messages instead of just enums, there is
no longer a need to combine callbacks with handlers

On Fri, 29 Jan 2016 at 19:01 Tomasz Masternak <notifications@github.com>
wrote:

> @andreasohlund <https://github.com/andreasohlund> we allow it currently.
> The question is do we still want to do it - is it really important. We were
> discussing that callback responses are inherently transient and it probably
> makes sense to have TTL on them either way.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/issues/3369#issuecomment-176651424>
> .
>
-- 

*John Simons*
*Engineer, Particular Software*

Email*:*  john.simons@particular.net
Web:    www.particular.net

**Comment by mauroservienti** [comment]
> there is no longer a need to combine callbacks with handlers

@johnsimons can you elaborate? we have this requirement:

> In case when callback has been registered and a handler for the same message type even if callback has been destroyed message will be processed by the handler.

**Comment by andreasohlund** [comment]
>no longer a need to combine callbacks with handlers

I can agree to that but still, what if user does it anyway?

There seems to be 2 types of replies, one where there is a handler waiting and one where there is a callback waiting.

Handler waiting: No need to be instance specific, arguably it should NOT be instance specific, since when scaled out all instances can process it. Should be durable since handlers means that it can survive a crash/restart

Callback waiting: Needs to go to a specific instance. Durability is not a must have

They seems to be conflicting?

**Comment by mauroservienti** [comment]
I don't see them conflicting, if:

* A handler is defined in an Endpoint it will be invoked
* If in an Endpoint instance there is also a callback registered the callback is invoked as well

That means that is an endpoint instance has registered a callback and the reply arrives both the handler and the callback are invoked. From the design perspective maybe there is no good reason to have both an handler and a callback for the same message type, I cannot think to any business scenario, but as I see it a registered callback is a short-living standard message handler.

**Comment by SzymonPobiega** [comment]
Here's how I see it: add this extensions to `SendOptions` and `ReplyOptions`:

```
public static void RouteReplyToThisInstance(this SendOptions option)
```

**Comment by andreasohlund** [comment]
>A handler is defined in an Endpoint it will be invoked

Unless there is a callback as well and the specific instance that requested the reply has been decommisioned? (isn't this a confict?)

**Comment by mauroservienti** [comment]
I'm not following, let me use a sample:

* Instance A send a message with a callback
    * Instance A has a message handler as well
* The reply comes in and Instance A is still alive
* The callback AND the message handler are invoked
    * exactly as when 2 message handlers are defined for the same message

Second scenario

* Instance A send a message with a callback
    * Instance A has a message handler as well
* The reply comes in and Instance A is not there anymore
* Another Instance N will pickup the message, only the message handler is invoked

**Comment by andreasohlund** [comment]
>The reply comes in and Instance A is not there anymore
Another Instance N will pickup the message, only the message handler is invoked

If we use instance specific queues for callbacks no other instance would pick the message up right? (so it would be "lost")

**Comment by mauroservienti** [comment]
yes, so who cares? As said to me is a design smell that you have both a callback and a message handler in the same endpoint, but if we treat a registered callback as a standard message handler a user could:

* create a standard message handler for all `IMessage`
* define in the configuration that the above message handler should run first
* register the callback

If the endpoint instance is not there anymore there is no good reason to invoke the generic message handler as well.

**Comment by mauroservienti** [comment]
Just had a quick chat with @andreasohlund and now I agree that combining a callback and a traditional message handler for the same message in the same endpoint makes no sense.

If a callback is registered for a message type and a message handler is defined for the same message type we should blow up preventing the user to do a thing that we do not support. This raises the issue that from the callbacks infrastructure we should be able to "query" core to gather the above information.

If a message handler is registered we need to be sure that the incoming message goes to the main Q where instances behave as competing consumers.

Did I miss something @andreasohlund?

**Comment by andreasohlund** [comment]
>Did I miss something @andreasohlund?

:+1:

**Comment by danielmarbach** [comment]
There is a good reason @mauroservienti. The handlers acts as a "catch all"
and you at least know from a code perspective that you received a reply
which was no longer handled by a callback. On a client you could still
visualize it in a generic way, track it in a database or whatever

**Comment by mauroservienti** [comment]
@danielmarbach I is the scenario I had in mind, but if the message is a reply to a callback and the instance is not there anymore there is no one looking at the callback queue so that message is lost anyway.

**Comment by SzymonPobiega** [comment]
@danielmarbach What @mauroservienti said :)

**Comment by SzymonPobiega** [comment]
I think it is OK for now to just now invoke the handlers. I believe in the long run we should expose callbacks in a special kind of integration endpoints where you can't host handlers.

**Comment by mauroservienti** [comment]
@SzymonPobiega are you thinking to SendOnly-callbacks enabled endpoints? :)

**Comment by SzymonPobiega** [comment]
@mauroservienti yes!

**Comment by danielmarbach** [comment]
@mauroservienti Only if the queues are volatile, then your statement is
true. Otherwise not ;)

**Comment by mauroservienti** [comment]
There is no one looking at the callback queue if the owning endpoint is gone, volatile or not.

**Comment by andreasohlund** [comment]
>Only if the queues are volatile

Or if you are dynamically scaling up and down? (think moving the slider on Azure)

**Comment by SzymonPobiega** [comment]
I've recorded a demo of the spike I mentioned with SQL transport (transport is unaware of any stuff related to instance ids and callbacks): https://drive.google.com/open?id=0B_QSSoVzhDxPMW1YaG1ycU5KTlE

**Comment by mauroservienti** [comment]
@tmasternak and I where asking ourselves if all what we are thinking here is backward compatible at the wire level, and if it must be.

**Comment by SeanFeldman** [comment]
I struggle with the idea of a handler for a callback reply as well. Feels like "reuse-abuse". Just because we can doesn't mean we should. 

TTL on a callback reply message/callback registration was identified as an orthogonal concern, but it feels as a part of this issue, not a separate one.

Should we go back to the requirement, and review it with the requirement owner?

**Comment by SzymonPobiega** [comment]
@SeanFeldman :+1: for going back to @udidahan with handlers for callbacks.

**Comment by mauroservienti** [comment]
Agree, :+1: for me as well

On Fri, Jan 29, 2016, 19:59 Szymon Pobiega <notifications@github.com> wrote:

> @SeanFeldman <https://github.com/SeanFeldman> [image: :+1:] for going
> back to @udidahan <https://github.com/udidahan> with handlers for
> callbacks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/issues/3369#issuecomment-176913485>
> .
>
-- 
Mauro Servienti | Solution Architect | Particular Software

**Comment by SzymonPobiega** [comment]
I assume @mauroservienti or @SeanFeldman are going to Udi to ask him about the handlers.

What's the next step with regards to `transport vs routing` problem?

**Comment by mauroservienti** [comment]
I'm off the whole week at a customer, @SeanFeldman can you arrange the details of how to involve @udidahan in the conversation?

What we need to discuss is:

* Time to live for callback messages
* Message handlers and callbacks in the same endpoint instance for the same message type, that as [stated here](https://github.com/Particular/NServiceBus/issues/3369#issuecomment-176683488) seems to be a no go

**Comment by SeanFeldman** [comment]
@mauroservienti wip, contacted @Chante-Vaturi to setup the call.

**Comment by SzymonPobiega** [comment]
@johnsimons are you good with proceeding with callbacks via routing for now?

**Comment by SzymonPobiega** [comment]
I've added a RabbitMQ WIP pull for my spike here: https://github.com/Particular/NServiceBus.RabbitMQ/pull/126

And a WIP pull for Callbacks package here: https://github.com/Particular/NServiceBus.Callbacks/pull/10

**Comment by mauroservienti** [comment]
@SzymonPobiega is routing our best (only?) option when it comes to callbacks backward compatibility?

**Comment by johnsimons** [comment]
commented on PR @SzymonPobiega

**Comment by SzymonPobiega** [comment]
When it comes to V5 -> V6 wire compat I thin it is. 

A V5 endpoint with callbacks enabled (SQL or RabbitMQ) will send a message to the shared queue adding a header like `NServiceBus.RabbitMQ.CallbackQueue`. V6 needs to use it instead of `reply-to` when the message intent is `reply`. That 'hack' can be implemented either on transport level or on core level, not sure which place is better.

**Comment by SzymonPobiega** [comment]
Guys, do you accept my proposal? The PR is here: https://github.com/Particular/NServiceBus/pull/3374 and the doco PR is here https://github.com/Particular/docs.particular.net/pull/1142

**Comment by andreasohlund** [comment]
:+1:

**Comment by SimonCropp** [comment]
closing this since https://github.com/Particular/NServiceBus/pull/3374 is merged

**Comment by SeanFeldman** [comment]
We're closing this issue, but were's a summary of what's in place for transports to implement callbacks?

**Comment by SimonCropp** [comment]
@SeanFeldman fair question. i will re-open until that is answered

**Comment by SzymonPobiega** [comment]
As #3374 is merged, *for now* the the callbacks are implemented in the core. That pull has TODO items I am working on to upgrade the downstreams (Callabacks, SQL Server and RabbitMQ).

There is a separate discussion going with regards to how do we generate the instance IDs.

**Comment by mauroservienti** [comment]
> There is a separate discussion going with regards to how do we generate the instance IDs.

where?

**Comment by SzymonPobiega** [comment]
@mauroservienti added the link.
